### ISSUE TITLE
scheduler_node: add preemption logic

### ISSUE THREAD

**Issue by jack-oquin** [utexas-bwi/concert_scheduling#27]
The current implementation never preempts resources it previously allocated, even when a higher-priority request is waiting on it.

**Comment by stonier** [comment]
Ok, time to open pandora's box :gift:

**Experimentation**

I did some playing around with the turtle teleop:

```
> rocon_launch turtle_concert turtle2.concert --screen
# in another shell
> rocon_remocon
```

In the remocon, select the user's role and the `Concert Teleop`. This lets you pre-empt a turtle from working for the concert `turtle_pond` service (running `turtle_concert/turtle_stroll.rapp`) and pull it in to work for the concert `teleop` service (running `turtle_concert/teleop`). 

This was all done by hacking the experimental `compatibility tree` scheduler of mine (it's seemingly more experimental by the day!). Here it makes a judgement call, deallocates the resource, stops the `turtle_stroll` rapp, reallocates it for `teleop` and starts the `turtle_concert/teleop` rapp.

There are of course, some major problems with this...and others lurking at the bottom of pandora's box.

**Grace Time**
- The scheduler shouldn't be immediately stopping and starting rapps.
- Services should be given a `grace time to allow them to gracefully release the resource from use (just in case it's in the middle of doing something).
- If a service doesn't release a resource within the grace time, the scheduler should then have permission to do something.

How does a requester release a resource? Currently there is the concept of cancelling requests, but not releasing individual resources.

**Pre-empting**
- Are we pre-empting a resource or a request ?

This question starts to get at the core of it I think. We could add alot of extra machinery for specific resource handling to preempt resources. The requester would need some handles to notify the scheduler it's `released` a resource from working in a current request. The scheduler would have to `deallocate` resources from requests if a grace timeout was reached. Actually the current method for handling clients that have gone is similar - we just flag a state on the resource and let the requester worry about an inadequately fulfilled current request. 

I think this is starting to sound quite complicated though. Should gone, or preempted resources be handled at the request level? i.e. requesters have to comply with cancelling an entire request within a grace time because one resource got preempted?

@jack-oquin this also brings me back to your earlier proposal about formulating requests with a single resource and then formulating requests of requests by applying certain logical operations. You could have a super-request that could flag subrequests as critical or not - i.e. a subrequest getting preempted could force cancelling of the whole request (critical) or subrequests could be granted permission to be cancelled themselves without cancelling the super-request. At the end of the day here, you are just dealing with requests either way.

@jack-oquin, @piyushk Lots of questions...no real firm idea on solution though so looking forward to whatever thoughts you have :)

**Comment by jack-oquin** [comment]
**Preemption**

> - Are we pre-empting a resource or a request ?

That is the fundamental issue:
- The resource pool tracks the CurrentStatus of known resources: {AVAILABLE, ALLOCATED, MISSING, GONE}.
- Request messages track the states of individual requests: {NEW, RESERVED, WAITING, GRANTED,  PREEMPTING, CANCELING, CLOSING}. 
- The fact that each Request can ask for multiple resources is somewhat of a conceptual mismatch. One would prefer those state changes to be synchronous.

For those reasons, I still like the idea of "element" requests that name a single resource and "sets" specifying some "union" or "intersection" of other requests. While that does not eliminate the complexity of preempting a single resource from a set union request, it does at least provide a clearer language for expressing the states that can arise.

But, we have not worked out the details of that proposal. My intuition about it could be wrong, and the effects of making such a fundamental change could be disruptive. You have a better idea than I do, when (if ever) that effort might be appropriate.

Unless we choose to make deep changes to the request structure, I believe we need to preempt at the request level. Requesters would need to understand that grouping several resources into a single request creates an "all or nothing" transaction. If more granularity is needed, they can be broken up into multiple requests, but we also understand the difficulties that introduces.

**Grace Time**

I would continue avoiding the grace period issue for a while longer by relying on cooperative preemption. I believe the necessary Request states already exist, although implementation might uncover some surprises.

That approach implies that the requester must cancel the entire request. It could simultaneously make a new request for what remains, but they would be separate items from the scheduler perspective.

**Comment by stonier** [comment]
Thanks for the reply Jack and sorry for the flatline on my end. Other events have caught up with me and I've had to postpone this for a short while. I'll pick up the thread again immediately following ROSKong.

**Comment by jack-oquin** [comment]
No problem. I've been busy releasing things and doing driver maintenance.

If I do find time to work on this any time soon, it will be something modest based on the current messages. The "pandora" option will await more detailed design discussions.
